### PR TITLE
Fixing SeaMonkey compatibility and improving few sites

### PR DESCRIPTION
**Pull request by UCyborg**
I hope someone finds these beneficial. Take it or leave it.

Edit: Actually, hold on. Reddit fix is not correct. One portion of the code further from regular expression has to be written differently. I'll do it later.

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -240,8 +261,10 @@ class RuleEngine {
                         this.selectors.set(selector, fixset);
                         accepted++;
                     }
-                } catch(e if e instanceof SyntaxError) {
-                    this._maybeRaiseSyntaxError(e, raiseErrors, selstr);
+                } catch (e) {
+                    if (e instanceof SyntaxError) {
+                      this._maybeRaiseSyntaxError(e, raiseErrors, selstr);
+                    }
```
Comment (martok) [inline lib/main.js:267]: How does SM not have the extended syntax? Wild.
Anyway, needs re-raising of non-matched exceptions.

File: lib/main.js [hunk 2]
```diff
@@ -281,8 +304,10 @@ class RuleEngine {
                     }
                 }
                 accepted ++;
-            } catch(e if e instanceof SyntaxError) {
-                this._maybeRaiseSyntaxError(e, raiseErrors, line);
+            } catch (e) {
+                if (e instanceof SyntaxError) {
+                  this._maybeRaiseSyntaxError(e, raiseErrors, line);
+                }
```
Comment (martok) [inline lib/main.js:310]: Same as above

File: lib/main.js [hunk 3]
```diff
@@ -611,8 +636,10 @@ function _optionsDisplayed(xul) {
             } else {
                 alert("Palefill", `Successfully parsed ${a}/${p} lines. Check debug output for details.`);
             }
-        } catch (e if e instanceof SyntaxError) {
-            alert("Palefill", e.message);
+        } catch (e) {
+            if (e instanceof SyntaxError) {
+              alert("Palefill", e.message);
+            }
```
Comment (martok) [inline lib/main.js:642]: ... and another one.

File: lib/main.js [hunk 4]
```diff
@@ -54,6 +58,14 @@ const _definitions = [
         selector: ["www.pixiv.net"],
         fix: ["std-customElements"]
     },
+    {
+        selector: ["www.redditstatic.com/desktop2x/CommentsPage.*.js$script"],
+        fix: ["$script", "reddit-comments-regexp"]
+    },
+    {
+        selector: ["www.youtube.com"],
+        fix: ["std-customElements"]
```
Comment (martok) [inline lib/main.js:67]: Youtube still deploys their own customElements Polyfill (all of webcomponents, actually), why is this required?
Adding this is much slower.

File: lib/polyfills.js [hunk 5]
```diff
@@ -144,6 +144,26 @@ exports.Array_flat = `if (typeof Array.prototype.flat !== "function") {
   };
 }`;
 
+exports.String_matchAll = `if (typeof String.prototype.matchAll !== "function") {
```
Comment (martok) [inline lib/polyfills.js:147]: Please add the source/copyright of this polyfill to the top of this file.

**Review by martok — CHANGES_REQUESTED** [review]
Hi @UCyborg!
Thank you for this PR! Seamonkey is something I can't test at all even if it is _technically_ supported, so I'm more than happy to get improvements for it.

However, there are a few issues with the current code that I'd like to see addressed first. Feel free to keep updating this PR (together with the reddit changes you mentioned) and we'll get this merged as one block.

**Comment by UCyborg** [comment]
Re-raising of non-matched exceptions: Correct me if I'm wrong, does this mean I just add "else throw e;" like shown [here](https://stackoverflow.com/questions/10486043/can-i-use-catche-if-e-instanceof-syntaxerror-in-node-jss-javascript)?

matchAll copyright: I think it's from SeaHOH (https://github.com/JustOff/github-wc-polyfill/commit/1b0e52569a430dabd9d7d680cbb6f4209e77a377), at least the code doesn't look similar compared to another popular implementation (https://github.com/es-shims/String.prototype.matchAll).

YouTube: You're right, window.customElements is actually defined with default user agent Pale Moon spoofs for it. Another funny thing, SeaMonkey says by default in its user agent that it's both Firefox 68 and SeaMonkey and then those preview thumbnails work (unlike with UA that just says Firefox 60), video description section has different appearance and customElements is also defined in this case by YouTube. Seems the only case when it isn't defined is with UA of Firefox 65 or newer.

I like the latter combination simply because then it seems to behave like it does on popular browsers. But yes, since the behavior is quirky, no problem, I'll throw that commit out and just use it on my locally installed copy.

BTW, is it OK if I just rebase and amend what's needed and force push to my repo? I'd prefer to have the history that isn't cluttered more than it has to be. I'm not concerned about someone forking my copy and getting confused later (don't think it'll come to that in this case), but will anything get messed up regarding this pull request?

**Comment by rofl0r** [comment]
> but will anything get messed up regarding this pull request?

no, a pull req is just getting updated after a force push with latest branch contents

**Comment by martok** [comment]
> Re-raising of non-matched exceptions: Correct me if I'm wrong, does this mean I just add "else throw e;" like shown [here](https://stackoverflow.com/questions/10486043/can-i-use-catche-if-e-instanceof-syntaxerror-in-node-jss-javascript)?

As far as I can tell, yes. I remember problems with wrong stack traces, but nobody mentions that anymore, so maybe that's a thing of the past.

> YouTube: You're right, window.customElements is actually defined with default user agent Pale Moon spoofs for it,  [...]
> I like the latter combination simply because then it seems to behave like it does on popular browsers. But yes, since the behavior is quirky, no problem, I'll throw that commit out and just use it on my locally installed copy.

Hm, interesting. I didn't take the spoofed UA into account. I guess a way to resolve that for the future could be to add a rule option for the user agent and use that to have rules that only activate on one browser.

Something odd: I have just tested your suggested change again and it _isn't_ that much slower anymore. When I last tested this about 3 months ago (just after the new design was rolled out), cold page load times for video pages went from 12 seconds to >20 seconds, that doesn't seem to be the case anymore.
I'll revisit that, but it needs a bit more testing to see what else might break without their polyfills. Leave it out of this one for now, please.

> BTW, is it OK if I just rebase and amend what's needed and force push to my repo? I'd prefer to have the history that isn't cluttered more than it has to be. I'm not concerned about someone forking my copy and getting confused later (don't think it'll come to that in this case), but will anything get messed up regarding this pull request?

As @rofl0r said, that just works. Github keeps a copy of the old review comments, but the PR itself always works with the current branch state.

**Comment by martok** [comment]
Perfect, thank you!

**Comment by UCyborg** [comment]
Looks like Stack Exchange updated their script so it doesn't have the incompatible code anymore.

**Comment by jchwdv** [comment]
Thank you both. You make my SeaMonkey usable with github again. Thank you :+1:

**Pull request merged**